### ISSUE TITLE
jsonConfig port check validation does not account for bind address

### ISSUE THREAD

**Issue by raintonr** [ioBroker/ioBroker.admin#2096]
Version 6.8.0
node: v16.20.2
js-controller: 5.0.8

When using jsonConfig element such as...

```json
"port": {
  "label": "Listen port",
  ...
},
```

.. automatic validation takes place, and if the given port number is in use by another adapter error, *"Port is already used by xxx"* is shown.

However, this does not account for bind addresses. For example, in my installation I want port 443 on an internal IP address to be used by admin.0 and port 443 on an external IP address to be used by another adapter. As it stands jsonConfig cannot accommodate this.

**Comment by foxriver76** [comment]
I guess we are talking about using `type: port`, as the element type is missing in the example.

**Comment by foxriver76** [comment]
Do I get it right, that you want to disable the scanning for a free port as you are also using it to specify the port of an external system? So, e.g. a flag to disable the checking would help you `external: boolean`?

**Comment by raintonr** [comment]
No. I like the scanning. Just it doesn't understand that if I have admin (for example) configured to use port 443 on bind address x.x.x.x then there is no reason why port 443 cannot be used in another adapter if that happens to bind address y.y.y.y.

**Comment by mcm1957** [comment]
If I read this correctly then the jsonConfig would require to know to which ip you later plan to bind the port. How should jsonConfig guess that you plan to bind this port to another ip address? How should this be specified in json? JsonConfig cannot know to which address you plan to bind later.

I too think that the only possibility would be to disable to check optionally.
Or that you do not use port but number input.

**Comment by raintonr** [comment]
How I'd build it:

1. If there is no bind field, or the bind field is empty or 'all interfaces' current behaviour.
2. If there is a bind field with valid address, only show a port error if that port and bind address combination are used in another adapter (currently only port is checked, not the combination).
3. When checking other adapters in step 2. be sure to account for 'all interfaces' (0.0.0.0 and ::).

This means when a user selects a port currently in use the error may appear, but if they then choose a unique bind address it should disappear.

**Comment by raintonr** [comment]
I have just encountered this error again, this time across hosts. Eg:

- Adapter 'X' running on host 'A' on port 8080.
- Adapter 'Y' running on host 'B' cannot be configured to use port 8080 because of the *"Port is already used by xxx"* message.

In this case you would hope accounting for bind address would solve this, but... in the case both hosts listen on their loopback address that would not be so. Hence port and bind address **AND** host the adapter is running on need to be checked before throwing this error.

**Comment by foxriver76** [comment]
Can you create a separate issue for respecting the host, this can be easily fixed.

**Comment by foxriver76** [comment]
I have created a PR for the different hostname problem https://github.com/ioBroker/ioBroker.admin/pull/2131

> If there is no bind field, or the bind field is empty or 'all interfaces' current behaviour.

So you would simply identify by a hardcoded property `adapter.native.bind`? The problem is, that adapter devs could also name this prop differently. But would maybe be a working heuristic.

**Comment by mcm1957** [comment]
I would not use a hardcoded property name. Why not add an attribute naming the bind attribute to use to the port field? Would ensure maximum compatibility too as when its missing the current behavior ccould be kept

**Comment by foxriver76** [comment]
yeah but still not ensures to catch all what if one adapter names it `bind` at the other `bindIp` or stuff like this, how would a dev know in advance?

**Comment by foxriver76** [comment]
The logic just reads `system.adapter.<instance>` objects and checks some props, i.e. is on same host, uses same port (also happens just by prop name unfortunately), so overall its all just a heuristic.

**Comment by raintonr** [comment]
> So you would simply identify by a hardcoded property `adapter.native.bind`? The problem is, that adapter devs could also name this prop differently.

Someone (not me) decided to check for an already utilised port based on some configuration property. You'd have to ask them what their logic behind that was :wink:
